src/util/kaldi-table-inl.h

```diff
@@ -245,104 +275,209 @@ template<class Holder>  class SequentialTableReaderScriptImpl:
-        state_ = kLoadFailed;
         return false;
+      } else {
+        if (holder_.Read(data_input_.Stream())) {
+          state_ = kHaveObject;
+        } else {  // holder_ will not contain data.
+          KALDI_WARN << "Failed to load object from "
+                     << PrintableRxfilename(data_rxfilename_);
+          return false;
+        }
       }
     }
+    // OK, at this point the state must be either
+    // kHaveObject or kHaveRange.
+    if (range_.empty()) {
+      // if range_ is the empty string, we should not be in the state
+      // kHaveRange.
+      KALDI_ASSERT(state_ == kHaveObject);
+      return true;
+    }
+    // range_ is nonempty.
+    if (state_ == kHaveRange) {
+      // range was already extracted, so there nothing to do.
+      return true;
+    }
+    // OK, range_ is nonempty and state_ is kHaveObject.  We attempt to extract
+    // the range object.  Note: ExtractRange() will throw with KALDI_ERR if the
+    // object type doesn't support ranges.
+    if (!range_holder_.ExtractRange(holder_, range_)) {
+      KALDI_WARN  << "Failed to load object from "
+                  << PrintableRxfilename(data_rxfilename_)
+                  << "[" << range_ << "]";
+      return false;
+    } else {
+      state_ = kHaveRange;
+      return true;
+    }
+  }
+
+  void SetErrorState() {
+    state_ = kError;
+    script_input_.Close();
+    data_input_.Close();
+    holder_.Clear();
+    range_holder_.Clear();
   }
 
   // Reads the next line in the script file.
+  // Possible entry states: kHaveObject, kHaveRange, kHaveScpLine, kFileStart.
+  // Possible exit states: kEof, kError, kHaveScpLine, kHaveObject.
   void NextScpLine() {
-    switch (state_) {
-      case kLoadSucceeded: holder_.Clear(); break;
-      case kHaveScpLine: case kLoadFailed: case kFileStart: break;
+    switch (state_) {  // Check and simplify the state.
+      case kHaveRange:
+        range_holder_.Clear();
+        state_ = kHaveObject;
+        break;
+      case kHaveScpLine: case kHaveObject: case kFileStart: break;
       default:
         // No other states are valid to call Next() from.
         KALDI_ERR << "Reading script file: Next called wrongly.";
     }
+    // at this point the state will be kHaveObject, kHaveScpLine, or kFileStart.
     std::string line;
     if (getline(script_input_.Stream(), line)) {
-      SplitStringOnFirstSpace(line, &key_, &data_rxfilename_);
-      if (!key_.empty() && !data_rxfilename_.empty()) {
+      // After extracting "key" from "line", we put the rest
+      // of "line" into "rest", and then extract data_rxfilename_
+      // (e.g. 1.ark:100) and possibly the range_ specifer
+      // (e.g. [1:2,2:10]) from "rest".
+      std::string data_rxfilename, rest;
+      SplitStringOnFirstSpace(line, &key_, &rest);
+      if (!key_.empty() && !rest.empty()) {
         // Got a valid line.
-        state_ = kHaveScpLine;
+        if (rest[rest.size()-1] == ']') {
```

Comment:
hmm it's a feature of c++11...
On Apr 3, 2016 7:39 PM, "Kirill Katsnelson" <notifications@github.com>
wrote:

> In src/util/kaldi-table-inl.h
> <https://github.com/kaldi-asr/kaldi/pull/647#discussion_r58319861>:
>
> >          // Got a valid line.
> > -        state_ = kHaveScpLine;
> > +        if (rest[rest.size()-1] == ']') {
>
> Use string::back()?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/647/files/1db3c939c2c250276f89f00ad80fcca3bf289d75#r58319861>
>